### PR TITLE
Filesystem usage analyzer

### LINKED ISSUE #3643: Add analyzer for `FileSystem`
## Problem
Without a strict enforcement mechanism, we risk a regression where file system access might bypass the `IFileSystem`.

## Context
Certain platforms, such as the Nintendo Switch, impose strict limitations on disk access. To support these platforms (i.e. allow the Unity SDK to fallback on preventing disk access OOTB) and prevent user projects from crashing, we introduced an option to opt-out of all disk-writing operations.

## Current Implementation
We now route file system access through our `IFileSystem` wrapper. Depending on the configuration in options, the SDK uses either a `ReadOnly` or `ReadWrite` filesystem. This mechanism acts as a fallback to prevent regressions. SDK developers must explicitly check `Options.FileWriteDisable` before attempting any disk write operation. If a write attempt is made with disk access disabled, the filesystem simply returns `false` without further context.

- [PR #3614](https://github.com/getsentry/sentry-dotnet/pull/3614)
- [PR #3641](https://github.com/getsentry/sentry-dotnet/pull/3641)

## Solution
We could/should create an analyzer to ensure that all file system access continues to be routed through the SDK’s custom wrappers. 
For reference on creating custom analyzers: [How to write a C# analyzer](https://learn.microsoft.com/en-us/dotnet/csharp/roslyn-sdk/tutorials/how-to-write-csharp-analyzer-code-fix)

### PR DESCRIPTION
**Pull request by RRode**
Resolves #3643

Implements analyzer that searches for file system usage and suggests IFileSystem usage. Currently checks for usages of File, FileInfo, Directory and DirectoryInfo. No warning is produced, if these are used within IFileSystem implementation.

Analyzer was added to Sentry project and finds usage of file system outside of IFileSystem implementations, which then breaks the build. Usages found in PortablePdbReader.cs, CFunctions.cs and FileDiagnosticLogger.cs.

@bitsandfoxes Is this roughly what you had in mind? 

#skip-changelog

### CODE REVIEW

**Comment by bruno-garcia** [comment]
This is really awesome and promising. Thank you!

Is there a way to ignore some places where we it would be safe to call from, for example the CFunction one perhaps will be

**Comment by RRode** [comment]
Is there a reason why you could not use `#pragma warning disable / enable`? IMO, with additional comment, this would make it explicit and well documented why the warning should be ignored. But, if you want, I could add some exceptions to the rule. Just let me know what you need.
Also is it OK that I placed the new projects under src and test? For me the analyzer project feels a bit odd there, since it's not a deliverable. I thought about putting both projects under modules, but that seemed a bit off too, since those are git submodules that get pulled in.
Other thing is, would it make sense to put the analyzer into Directory.Build.props in src, so it is added in all src files? I assume that you want to enforce the warning in those projects as well.

**Comment by bruno-garcia** [comment]
> Is there a reason why you could not use #pragma warning disable / enable? IMO, with additional comment, this would make it explicit and well documented why the warning should be ignored. But, if you want, I could add some exceptions to the rule. Just let me know what you need.

That makes sense, `#pragma` sounds like the right solution

Also is it OK that I placed the new projects under src and test? For me the analyzer project feels a bit odd there, since it's not a deliverable. I thought about putting both projects under modules, but that seemed a bit off too, since those are git submodules that get pulled in.

I also think it's kind of odd, but it's not a big deal. Alternatively it could be a new top level directory: `analyzers` or `build-src` I dunno. I think as you did it's fine. But worth adding a README.md on the project directory to clarify this analyizer is only for consumption from `Sentry` itself, for development of the SDK.

Other thing is, would it make sense to put the analyzer into Directory.Build.props in src, so it is added in all src files? I assume that you want to enforce the warning in those projects as well.

We only really need this on `Sentry` since this is a requirement coming from `Sentry.Unity`. So, technically we'd want that analyizer there too.

We'd include that here too: https://github.com/getsentry/sentry-unity/tree/main/src/Sentry.Unity
Using relative paths to the sentry-dotnet submodule

File: src/Sentry.Analyzers/Resources.resx
Comment (bruno-garcia) [inline src/Sentry.Analyzers/Resources.resx:22]: If it's optional, can we got without them? If not, I suggest taking more details from the ticket that introduced this requirement (tldr: Switch file system requirements)

```suggestion
    <value>Route file system access through IFileSystem wrapper.</value>
```

File: src/Sentry/Sentry.csproj
Comment (bruno-garcia) [inline src/Sentry/Sentry.csproj:53]: Does this need `PrivateAssets>all` ?

Comment (RRode) [reply src/Sentry/Sentry.csproj]: If I understand the [docs](https://learn.microsoft.com/en-us/nuget/consume-packages/package-references-in-project-files#controlling-dependency-assets) correctly, it could have even been fine, but I added it anyway. I also added private assets in analyzer references.

File: src/Sentry.Analyzers/Resources.resx
Comment (RRode) [reply src/Sentry.Analyzers/Resources.resx]: I completely dropped the resx and hardcoded the warning, since I assume that no warning translations will be needed.

**Comment by RRode** [comment]
> That makes sense, `#pragma` sounds like the right solution

Can you provide some comments or should I just disable the warning in currently offending files?

> I also think it's kind of odd, but it's not a big deal. Alternatively it could be a new top level directory: `analyzers` or `build-src` I dunno. I think as you did it's fine. But worth adding a README.md on the project directory to clarify this analyizer is only for consumption from `Sentry` itself, for development of the SDK.

Then I will keep it simple for now and leave it as it is. I added the README, but was not able to remove the one added via Directory.Build.props to the project, so I hope that is OK. 

> We only really need this on `Sentry` since this is a requirement coming from `Sentry.Unity`. So, technically we'd want that analyizer there too.
>
> We'd include that here too: https://github.com/getsentry/sentry-unity/tree/main/src/Sentry.Unity
Using relative paths to the sentry-dotnet submodule

I guess it makes sense to first merge it here and then add it in the other repo, right?

File: src/Sentry.Analyzers/Sentry.Analyzers.csproj
Comment (bruno-garcia) [inline src/Sentry.Analyzers/Sentry.Analyzers.csproj:13]: afaik these default to the project name so are redundant 
```suggestion
```

**Review by bruno-garcia — COMMENTED** [review]
Looks good. 

Doesn't seem to be draft anymore, ready for a final pass and merge?

**Comment by bruno-garcia** [comment]
> Can you provide some comments or should I just disable the warning in currently offending files?

The current ones can be skipped, thanks

> Then I will keep it simple for now and leave it as it is. I added the README, but was not able to remove the one added via Directory.Build.props to the project, so I hope that is OK.

Sounds good

> I guess it makes sense to first merge it here and then add it in the other repo, right?

Definitely, that was just a note for reference, doesn't affect this PR. Can be a follow up if you'd like to contribute further 🙏

File: src/Sentry.Analyzers/Sentry.Analyzers.csproj
Comment (RRode) [reply src/Sentry.Analyzers/Sentry.Analyzers.csproj]: I removed them

**Comment by RRode** [comment]
> The current ones can be skipped, thanks

I added the pragmas in all the places, but this requires changes to [this submodule](https://github.com/getsentry/Ben.Demystifier/pull/6). If I understand correctly this needs to be merged first to main and then reference in this PR adjusted? I am not used to this submodule workflow, so I hope I am doing it right. 😅
 
> Definitely, that was just a note for reference, doesn't affect this PR. Can be a follow up if you'd like to contribute further 🙏

I will give it a try. 🙂

File: src/Sentry/Infrastructure/FileDiagnosticLogger.cs
Comment (bruno-garcia) [inline src/Sentry/Infrastructure/FileDiagnosticLogger.cs:34]: Ah good point, @bitsandfoxes anything we need to do to make this "Not available on Switch" ? 

Edge case since someoen needs to get out of their way to use this and its' just for debugging, but this analyzier really help us make these calls now.

Not doing anything about this (just ignoring it) is fine by me too but we should add a note to the docs that this thing isn't supported on Switch anyway:

https://docs.sentry.io/platforms/unity/configuration/diagnostic-logger/#FileDiagnosticLogger

File: src/Sentry/Platforms/Native/CFunctions.cs
Comment (bruno-garcia) [inline src/Sentry/Platforms/Native/CFunctions.cs:161]:
```suggestion
            // Allow direct file system usage - Not application for Nintendo Switch
```

**Review by bruno-garcia — APPROVED** [review]
I love this! thank you so much

**Comment by bruno-garcia** [comment]
I'll leave it for @jamescrosswell to do the merging as he's the maintainer but LGTM! 🙏

**Comment by jamescrosswell** [comment]
> I added the pragmas in all the places, but this requires changes to [this submodule](https://github.com/getsentry/Ben.Demystifier/pull/6). If I understand correctly this needs to be merged first to main and then reference in this PR adjusted? I am not used to this submodule workflow, so I hope I am doing it right. 😅

Hey @RRode, up until now we've been maintaining the Sentry fork of Ben.Demystifier purely for use in the sentry-dotnet repository so the workflows for changing code in there are a bit obscure and not very contributor friendly (the code used by `sentry-dotnet` isn't in the `main` branch, for exaple). I'll do a bit of head scratching to see if we could make this easier.

Worst case, I can cherry pick the change from your PR and get it in there but I'd rather tidy up the submodule so it's easier to accept contributions. Give me a couple days.

**Comment by jamescrosswell** [comment]
> Worst case, I can cherry pick the change from your PR and get it in there but I'd rather tidy up the submodule so it's easier to accept contributions. Give me a couple days.

Hi @RRode, I ended up just cherry pick your changes and push these up.

**Comment by jamescrosswell** [comment]
> @jamescrosswell I pushed the missing warning disable for AndroidHelpers

Thanks @RRode - apologies in advance if the build breaks on this one. GitHub just changed their macos-14 runners (no longer support XCode 16.0) so we need to bump to macos-15 but that's missing a bunch of stuff we need. Once this is merged, we can pull the changes into this branch and the build should work again:
- https://github.com/getsentry/sentry-dotnet/pull/3754

**Comment by RRode** [comment]
@jamescrosswell no problem. What about this format code check? Is there some doing from my side? If I understand this correctly, git fails to do something with my branch.

**Comment by jamescrosswell** [comment]
> @jamescrosswell no problem. What about this format code check? Is there some doing from my side? If I understand this correctly, git fails to do something with my branch.

Hm, possibly it's getting confused because there are multiple remotes defined for the Ben.Demystifier submodule (getsentry/Ben.Demystifier and benadams/Ben.Demystifier). The benadams (upstream) seems to be dead - it hasn't been updated forever and we can't get in touch with Ben so maybe we just remove that remote. 

I'll look into it once we've fixed the macos builds.

**Comment by jamescrosswell** [comment]
> @jamescrosswell no problem. What about this format code check? Is there some doing from my side? If I understand this correctly, git fails to do something with my branch.

Hm, actually I think it might be an issue with `scripts/commit-formatted-code.sh`:
https://github.com/getsentry/sentry-dotnet/blob/133e1a5fce1c80a43c79505e2c9f534a118ae6b0/scripts/commit-formatted-code.sh#L13

In your case, the PR is coming from a fork so it's not finding the `filesystem-usage-analyzer` branch (it needs to checkout `RRode:filesystem-usage-analyzer`).

I'm in GMT+12 so it's kind of late on a Friday night here, but I'll look into it on the weekend or early next week.

**Comment by jamescrosswell** [comment]
Finally got it over the line. Thanks for your patience and huge thank you for the PR @RRode !

**Comment by RRode** [comment]
Happy to help! 🙂

**Pull request merged**
